### DISCUSSION TITLE
useFieldArray strips keyName from object

### DISCUSSION THREAD

**Discussion by steveoh** [react-hook-form/react-hook-form#6621] [Q&A] [answered]
Since `useFieldArray` does not work on flat arrays, I would like to use the most simple object format which would be `{id}` (_since keyName defaults to id_) but when a form is submitted with a `useFieldArray` item with only an `id` property, the `keyName` field is removed from the object. Therefore I have to duplicate my id and have use an object like `{id, key}` to be available on form submission. For what reason is the `keyName` stripped from the object on form submission?

**Comment by bluebill1049** [comment, marked as answer]
you can rename `keyName` to other
```
useFieldArray({
  keyName: 'whatever'
})
<input key={field.whatever} />
```

because the key itself is only made use for the rendering only, it's not part of the form value.

**Reply by steveoh** [reply]
for my instance, setValue does not allow me to set the value of a virtual input to an array so I migrated to setFieldArray which does. I think your assumption that the key is only used for rendering is assuming too much. Would it be a breaking change to not strip this value? 

I use key properties on react lists all the time that are unique but are also used for other things besides rendering.

Does that make sense?

**Reply by bluebill1049** [reply]
> I think your assumption that the key is only used for rendering is assuming too much.

this is not an assumption this is made so it works with uncontrolled inputs. (by design)

> Would it be a breaking change to not strip this value?

100% breaking change

so my suggested solution is not a workaround? to have your own id input? and change the keyName to other?

**Reply by steveoh** [reply]
Ok I understand now. To take a step back, am I using this hook properly to store a single array value for a virtual input? It seems the intended use case is different.

**Reply by bluebill1049** [reply]
what about this? (haven't test it)

```tsx
const { fields } = useFieldArray({
  name: 'test',
  keyName: 'key',
})

{fields.map((field, index) => {
  <Controller key={field.key} name={'test.${index}.id'} render={() => null } />
})}
```

**Reply by adriantrunzo** [reply]
Hi all. I am moving my discussion started in [this issue](https://github.com/react-hook-form/react-hook-form/issues/6639) here since it seems more appropriate. Sorry, I didn't find this discussion before opening the issue.

@bluebill1049 I think what you are saying is the following:
```
You should not have a form input (rendered or virtual) whose name matches the value of keyName when using a field array.
```

Is that statement correct? If so, I have a few suggestions/questions:
- The documentation could be updated to make it clear that you need to change `keyName` if it matches a property name in your form data. Right now, the importance of using the keyName property (eg `id`) as the `key` prop is well described, but there is no detail I can find in the docs about `keyName` matching an input name.
- It would be helpful if the library showed a warning if a property `[keyName]` is already defined on the field data. For example, with the following setup:
```jsx
const { control } = useForm({
  defaultValues: {
    myarray: [
      { id: 1, name: 'Test' },
      { id: 2, name: 'Example' }
    ]
  }
})

const { fields } = useFieldArray({
  name: 'myarray',
  // Default keyName id
})
```
The library could show a warning like "A property matching the keyName id is already defined on the field data, please choose a different keyName."
- I realize it would be a breaking change, but maybe use a different default keyName than `id`? `id` seems very common and likely to conflict with peoples' data.

Thanks for the discussion.

**Reply by TimHav** [reply]
@bluebill1049 I agree with @adriantrunzo that the docs could be clearer (I am still on v6 myself). Reading the docs, I assumed that the `keyName` should match a unique property on my data. But the opposite is true: it should not match any property on my data.
